### PR TITLE
ref(ingest): Create an arroyo-based ingest consumer

### PR DESCRIPTION
**Pull request by Swatinem**
Adds a simplistic arroyo-based variant of the ingest consumers. This consumer is fully sequential without batching its messages, and it delegates to the same functions as the former consumer did.

To enable the new variant, pass `--v2-consumer` as argument to the CLI. The new consumer no longer supports listening to multiple topics, so each of the three ingest consumers must be started separately with this flag enabled. Do not enable parallelism, as this may lead to data inconsistencies.

In a follow-up, the flag will be removed and arroyo becomes the default.

### CODE REVIEW

File: src/sentry/ingest/consumer_v2/ingest.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+    raw_payload = message.payload.value
+    message: DecodedMessage = msgpack.unpackb(raw_payload, use_list=False)
+
+    message_type = message["type"]
+    project_id = message["project_id"]
+
+    # TODO: we are only fetching a single project, so should we use a different metric?
+    with metrics.timer("ingest_consumer.fetch_projects"):
+        project = Project.objects.get_from_cache(id=project_id)
```
Comment (jan-auer) [inline src/sentry/ingest/consumer_v2/ingest.py:45]: Fetching single projects at a time adds quite a bit of latency. We should test this (as this is the most straight forward implementation). If this turns out to be too large in latency, then we can instead:
- collect messages into batches
- add a task step that runs `get_from_cache_many`
- add a multiprocessing task that does the rest

File: src/sentry/ingest/consumer_v2/ingest.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        if result is not None:
+            data, callback = result
+            callback(_store_event(data))
+
+    elif message_type == "attachment_chunk":
+        process_attachment_chunk(message, projects)
+    elif message_type == "attachment":
+        process_individual_attachment(message, projects)
+    elif message_type == "user_report":
+        process_userreport(message, projects)
```
Comment (jan-auer) [inline src/sentry/ingest/consumer_v2/ingest.py:59]: I'd disable this code and raise an exception instead. Let's get this consumer working for just events and transactions first. 

Streaming team also suggested that we move away from multiplexing consumers. In this case we can have a simplified version for profiles, events, and transactions (without attachments) that just deals with event payloads. That's the one I'd test here.

File: src/sentry/ingest/consumer_v2/ingest.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    else:
+        raise ValueError(f"Unknown message type: {message_type}")
+
+    # TODO: this is not `flush` anymore, so should we use a different metric here?
```
Comment (jan-auer) [inline src/sentry/ingest/consumer_v2/ingest.py:63]: Theres a builtin metric in arroyo now somewhere, so this can be removed.

File: src/sentry/ingest/consumer_v2/ingest.py (hunk 1 above)
Comment (Swatinem) [reply src/sentry/ingest/consumer_v2/ingest.py:45]: If this is only about latency, I don’t think that matters too much, as at least events will be parallelized. Maybe even going through all the ceremony of batching things might add more latency than that.

File: src/sentry/utils/batching_kafka_consumer.py [hunk 4]
```diff
@@ -215,6 +215,7 @@ def __init__(
         metrics_sample_rates=None,
         metrics_default_tags=None,
         commit_on_shutdown: bool = False,
+        **_kwargs,
```
Comment (untitaker) [inline src/sentry/utils/batching_kafka_consumer.py:218]: why is this necessary?

Comment (Swatinem) [reply src/sentry/utils/batching_kafka_consumer.py:218]: The v1 and v2 consumers have different options which will just be passed down further as kwargs, in order to avoid errors because of this, this just accepts anything that was passed from higher up.
I prefer that to explicitly `pop`-ing off any additional argument, as that will break when you forget to do so.

File: src/sentry/ingest/consumer_v2/factory.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        force_cluster=force_cluster,
+    )
+    consumer = KafkaConsumer(consumer_config)
+
+    # The `attachments` topic that is used for "complex" events needs ordering
+    # guarantees: Attachments have to be written before the event using them
+    # is being processed. We will use a simple serial `RunTask` for those
+    # for now.
+    # For all other topics, we can use multi processing.
+    allow_multi_processing = topic != settings.KAFKA_INGEST_ATTACHMENTS
```
[the remark this answers is not in the document]
Comment (Swatinem) [reply src/sentry/ingest/consumer_v2/factory.py:96]: good catch with the mixed-up if condition. Otherwise I would like to keep the check, which will eventually go away with the followup PR.

File: tests/sentry/ingest/ingest_consumer/test_ingest_consumer_kafka.py [hunk 6]
```diff
@@ -105,14 +111,31 @@ def test_ingest_consumer_reads_from_topic_and_calls_celery_task(
...
-        consumer = get_ingest_consumer(
-            max_batch_size=2,
-            max_batch_time=5000,
-            group_id=random_group_id,
-            consumer_types={ConsumerType.Events},
-            auto_offset_reset="earliest",
-            executor=executor,
-        )
+        if ingest_v2:
+            DEFAULT_BLOCK_SIZE = int(32 * 1e6)  # whatever this even means
```
[the remark this answers is not in the document]
Comment (Swatinem) [reply tests/sentry/ingest/ingest_consumer/test_ingest_consumer_kafka.py:115]: thanks for the explanation :-) IMO the constructor has way too many tuneables, and some sensible defaults would be nice.

File: src/sentry/ingest/consumer_v2/factory.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+def get_config(
+    topic: str,
+    group_id: str,
+    auto_offset_reset: str,
+    strict_offset_reset: bool,
+    force_cluster: str | None,
+) -> MutableMapping[str, Any]:
+    cluster_name: str = force_cluster or settings.KAFKA_TOPICS[topic]["cluster"]
+    create_topics(cluster_name, [topic])
```
Comment (lynnagara) [inline src/sentry/ingest/consumer_v2/factory.py:125]: I don't think you need this line, all the topics are created in the devserver now

File: src/sentry/runner/commands/run.py [hunk 8]
```diff
@@ -540,6 +551,30 @@ def ingest_consumer(consumer_types, all_consumer_types, **options):
...
+                "Only a single `--consumer-type` allowed when using the `v2` ingest consumer"
+            )
+
+        # FIXME: Judging by the default values, `max_batch_time` is in milliseconds,
+        # but arroyo expects that option to be in seconds, so fix that up here:
+        if (mbt := options["max_batch_time"]) > 10:
+            options["max_batch_time"] = mbt / 1000
+
+        topic = ConsumerType.get_topic_name(next(iter(consumer_types)))
+        consumer = get_ingest_consumer(topic=topic, **options)
```
Comment (lynnagara) [inline src/sentry/runner/commands/run.py:568]: You need to call `configure_metrics()` with a metrics backend before this in order for Arroyo to emit any metrics https://getsentry.github.io/arroyo/metrics.html

File: src/sentry/utils/batching_kafka_consumer.py (hunk 4 above)
Comment (untitaker) [inline src/sentry/utils/batching_kafka_consumer.py:218]: Please do not do this. We had an issue where we silently ignored CLI arguments because of "convenience" changes like this, and it wasted a bunch of time when we tried to tune those CLI parameters that had no effect, and it took multiple days to fix this for all consumers.

**Review by untitaker — CHANGES_REQUESTED** [review]

File: src/sentry/utils/batching_kafka_consumer.py (hunk 4 above)
Comment (jan-auer) [reply src/sentry/utils/batching_kafka_consumer.py:218]: This was already brought up before and dismissed by the author: https://github.com/getsentry/sentry/pull/49957#discussion_r1214167717

It seems the intent is compatibility rather than convenience and this will be removed when the migration is complete -- would you prefer we list all options and selectively pass them instead?

File: src/sentry/runner/commands/run.py (hunk 8 above)
Comment (jan-auer) [reply src/sentry/runner/commands/run.py]: Added in 18c0270

File: src/sentry/utils/batching_kafka_consumer.py (hunk 4 above)
Comment (untitaker) [reply src/sentry/utils/batching_kafka_consumer.py:218]: I did read the response on that comment. I am concerned about this making us silently (and accidentally) ignore CLI parameters that were supposed to be honored, because that's exactly what had already happened in this past (I think it was literally this function that already silently discarded kwargs), and for the exact same stated reason.

If that were only affecting the legacy ingest consumer, I wouldn't care since it will probably be deleted in the next weeks. But batching-kafka-consumer is used by more consumers than that.

I would instead explicitly `options.pop(..)` all unsupported kwargs in `runner/commands/run.py` around line 572. While this still leaves over a bunch of ignored CLI parameters (block size), CLI declaration and list of ignored options are now at least in the same file.

I care about this because this quarter we have an initiative planned (not started) where we attempt to unify all this CLI boilerplate + metrics across consumers, so they work a little bit more like Snuba consumers in their operation.

File: src/sentry/runner/commands/run.py (hunk 8 above)
Comment (untitaker) [reply src/sentry/runner/commands/run.py]: now the tests are failing because `configure_metrics` leaks from previous tests. we have no precedent for how to solve this problem, because the arroyo consumers typically test processing strategies instead of entire CLI commands. I would suggest to mock out `configure_metrics` in tests to noop...

**Review by untitaker — APPROVED** [review]

**Review by jan-auer — APPROVED** [review]
We just need to make sure that we do not enable multi-processing in this current state.
